File: typescript/src/resources/pii.ts
export class EventsPii {
  private _request: (method: string, path: string, body?: unknown) => Promise<unknown>;

  constructor(request: (method: string, path: string, body?: unknown) => Promise<unknown>) {
    this._request = request;
  }

  async create(sessionId: string, eventId: string, params: { pii_data: string }): Promise<{ message: string }> {
    return this._request("POST", `/sessions/${sessionId}/events/${eventId}/pii`, params) as Promise<{ message: string }>;
  }

  async retrieve(sessionId: string, eventId: string): Promise<{ pii_data: string }> {
    return this._request("GET", `/sessions/${sessionId}/events/${eventId}/pii`) as Promise<{ pii_data: string }>;
  }
}

export class SessionsPii {
  private _request: (method: string, path: string, body?: unknown) => Promise<unknown>;

  constructor(request: (method: string, path: string, body?: unknown) => Promise<unknown>) {
    this._request = request;
  }

  /**
   * Erase all PII records for a session (GDPR right to erasure).
   *
   * **Irreversible.** PII payloads cannot be recovered after deletion.
   * Hash chain integrity is preserved — only the PII payloads are deleted.
   */
  async delete(sessionId: string): Promise<{ message: string; records_erased: number; session_id: string }> {
    return this._request("DELETE", `/sessions/${sessionId}/pii`) as Promise<{
      message: string;
      records_erased: number;
      session_id: string;
    }>;
  }
}

File: python/shield/resources/pii.py
from typing import Any, Dict


class EventsPii:
    """PII management for session events."""

    def __init__(self, client):
        self._client = client

    def create(self, session_id: str, event_id: str, pii_data: str) -> Dict[str, Any]:
        """Store encrypted PII data for an event.

        Args:
            session_id: The session ID.
            event_id: The event ID.
            pii_data: Raw PII string to encrypt and store (max 512 KB).

        Returns:
            Dict with ``message`` confirmation.
        """
        return self._client._request(
            "POST",
            f"/sessions/{session_id}/events/{event_id}/pii",
            json_data={"pii_data": pii_data},
        )

    def retrieve(self, session_id: str, event_id: str) -> Dict[str, Any]:
        """Retrieve decrypted PII data for an event.

        Args:
            session_id: The session ID.
            event_id: The event ID.

        Returns:
            Dict with ``pii_data`` field.
        """
        return self._client._request(
            "GET",
            f"/sessions/{session_id}/events/{event_id}/pii",
        )


class SessionsPii:
    """Session-level PII erasure."""

    def __init__(self, client):
        self._client = client

    def delete(self, session_id: str) -> Dict[str, Any]:
        """Erase all PII records for a session (GDPR right to erasure).

        **Irreversible.** PII payloads cannot be recovered after deletion.
        Hash chain integrity is preserved — only the PII payloads are deleted.

        Args:
            session_id: The session ID.

        Returns:
            Dict with ``message``, ``records_erased``, and ``session_id``.
        """
        return self._client._request(
            "DELETE",
            f"/sessions/{session_id}/pii",
        )
